mono/sgen/sgen-gc.c

```diff
@@ -2242,64 +2243,49 @@ sgen_perform_collection (size_t requested_size, int generation_to_collect, const
 
 	TV_GETTIME (gc_total_start);
 
-	if (concurrent_collection_in_progress) {
-		/*
-		 * If the concurrent worker is finished or we are asked to do a major collection
-		 * then we finish the concurrent collection.
-		 */
-		gboolean finish = major_should_finish_concurrent_collection () || generation_to_collect == GENERATION_OLD;
-
-		if (finish) {
-			major_finish_concurrent_collection (wait_to_finish);
-			oldest_generation_collected = GENERATION_OLD;
-		} else {
-			SGEN_ASSERT (0, generation_to_collect == GENERATION_NURSERY, "Why aren't we finishing the concurrent collection?");
+	// FIXME: extract overflow reason
+	// FIXME: minor overflow for concurrent case
+	if (generation_to_collect == GENERATION_NURSERY && !finish_concurrent) {
+		/* We need a minor collection */
+		if (concurrent_collection_in_progress)
 			major_update_concurrent_collection ();
-			collect_nursery (NULL, FALSE);
-		}
-
-		goto done;
-	}
-
-	SGEN_ASSERT (0, !concurrent_collection_in_progress, "Why did this not get handled above?");
 
-	/*
-	 * There's no concurrent collection in progress.  Collect the generation we're asked
-	 * to collect.  If the major collector is concurrent and we're not forced to wait,
-	 * start a concurrent collection.
-	 */
-	// FIXME: extract overflow reason
-	if (generation_to_collect == GENERATION_NURSERY) {
-		if (collect_nursery (NULL, FALSE)) {
+		if (collect_nursery (NULL, FALSE) && !concurrent_collection_in_progress) {
 			overflow_generation_to_collect = GENERATION_OLD;
 			overflow_reason = "Minor overflow";
 		}
 	} else {
-		if (major_collector.is_concurrent && !wait_to_finish) {
-			collect_nursery (NULL, FALSE);
-			major_start_concurrent_collection (reason);
-			// FIXME: set infos[0] properly
-			goto done;
-		}
-
-		if (major_do_collection (reason, wait_to_finish)) {
-			overflow_generation_to_collect = GENERATION_NURSERY;
-			overflow_reason = "Excessive pinning";
+		/* We need a major collection */
+		if (finish_concurrent) {
```

Comment:
Could we please do this as an `else if (finish_concurrent)`?  That would not only reduce indentation but also make the diff smaller (and the code more readable).